### PR TITLE
Schema changing

### PR DESCRIPTION
**Pull request by Mil-m**
changing schemas from public to private and vice versa inside endpoints

### CODE REVIEW

File: kooplex-proxy/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (davidyuyuan) [inline kooplex-proxy/deploy.sh:1]: As per discussion, please move all test scripts to be under "tests".

File: kooplex-proxy/pod_exec.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Comment (davidyuyuan) [inline kooplex-proxy/pod_exec.sh:1]: As per discussion, please move all test scripts to be under "tests".

File: kooplex-proxy/proxy/crud.py [hunk 3]
```diff
@@ -1,33 +1,74 @@
 from sqlalchemy.orm import Session
-from sqlalchemy import text
+from sqlalchemy.ext.declarative import DeclarativeMeta
 
 import models
-from database import engine
+from database import SCHEMA_SET
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/crud.py:5]: As I asked in the in the previous PR, please change the data type to dictionary, where the keys are the nicknames of the schemas. Please discuss with Krizstian to agree on the keys.

File: kooplex-proxy/proxy/main.py [hunk 4]
```diff
@@ -27,42 +27,42 @@ def read_root():
 
 
 @app.get("/country_samples/", response_model=List[schemas.MView_CountrySamples])
-def read_country_samples(skip: int = 0, limit: int = 100):
-    country_samples = crud.get_country_samples(next(get_db()), skip=skip, limit=limit)
+def read_country_samples(skip: int = 0, limit: int = 100, schema='sandbox_public'):
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/main.py:30]: It is not a good idea to define a default schema name in functions. The schema names may likely different between the staging database and production database or from release to release.

File: kooplex-proxy/proxy/models.py [hunk 5]
```diff
@@ -1,31 +1,46 @@
...
     n_sample = Column(INTEGER)
     log_n_sample = Column(DOUBLE_PRECISION)
     relative_n_sample = Column(DOUBLE_PRECISION)
     relative_log_n_sample = Column(DOUBLE_PRECISION)
 
 
-class MViewHumanMetaMv(Base):
+class MViewHumanMetaMv(AbstractBase):
     __tablename__ = "app_human_meta_mv"
+    __table_args__ = {'schema': 'sandbox_public'}
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/models.py:33]: Again, it is a bad idea to populate hardcoded schema name everywhere. You should keep the default schema name in one place, perhaps, in the parent class.

File: kooplex-proxy/proxy/crud.py [hunk 6]
```diff
@@ -1,33 +1,74 @@
...
-    return db.query(models.MViewCountrySamples).offset(skip).limit(limit).all()
+def schema_changing(model: DeclarativeMeta, endp_schema: str):
+    current_schema = model.get_schema()
+    if current_schema != endp_schema:
+        if endp_schema in SCHEMA_SET:
+            model.set_schema(schema_name=endp_schema)
+            print(f"-- Changing schema from '{current_schema}' to '{endp_schema}' for the endpoint: country_samples --")
+            return 0
+        else:
+            print(f"This schema is unavailable for using, please use one of these schemas: {SCHEMA_SET}")
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/crud.py:16]: security exposure. You should only provide the nicknames / AKA dictionary keys in the error message.

File: kooplex-proxy/proxy/crud.py [hunk 7]
```diff
@@ -1,33 +1,74 @@
...
+            return 1
+    return 0
 
 
-def get_human_meta_mv(db: Session, skip: int = 0, limit: int = 100):
-    return db.query(models.MViewHumanMetaMv).offset(skip).limit(limit).all()
+def get_country_samples(db: Session, skip: int = 0, limit: int = 100, endp_schema='sandbox_public'):
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/crud.py:21]: Please do not hardcode schema names in the function parameters.

**Review by davidyuyuan — CHANGES_REQUESTED** [review]

File: kooplex-proxy/proxy/models.py [hunk 8]
```diff
@@ -1,31 +1,48 @@
...
+        return cls.__table_args__['schema']
+
+    @classmethod
+    def set_schema(cls, schema_name):
+        cls.__table_args__['schema'] = schema_name
+
+
+class MViewCountrySamples(AbstractBase):
     __tablename__ = "app_country_samples_full"
+    __table_args__ = {'schema': ALLOWED_SCHEMAS['schema_1']}
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/models.py:25]: I was thinking that this line "__table_args__ = {'schema': ALLOWED_SCHEMAS['schema_1']}" could be in the parent class so it gets inherited by child classes.

File: kooplex-proxy/proxy/crud.py [hunk 9]
```diff
@@ -1,33 +1,79 @@
...
+            model.set_schema(schema_name=endp_schema)
+            print(f"-- Changing schema from '{current_schema}' to '{endp_schema}' for the endpoint: {endpoint_name} --")
+            return 0
+        else:
+            return 0
 
 
-def get_human_meta_mv(db: Session, skip: int = 0, limit: int = 100):
-    return db.query(models.MViewHumanMetaMv).offset(skip).limit(limit).all()
+def get_country_samples(db: Session, skip: int = 0, limit: int = 100, endp_schema_key=''):
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/crud.py:26]: You did the right thing on line 8 "endp_schema_key: str". What is the reason to do it differently on line 26 and all the other functions? (e.g. endp_schema_key='')

**Review by davidyuyuan — CHANGES_REQUESTED** [review]

File: kooplex-proxy/proxy/models.py (hunk 8 above)
Comment (Mil-m) [reply kooplex-proxy/proxy/models.py:25]: yes, indeed

File: kooplex-proxy/proxy/crud.py (hunk 9 above)
Comment (Mil-m) [reply kooplex-proxy/proxy/crud.py:26]: here
def schema_changing(model: DeclarativeMeta, endpoint_name: str, endp_schema_key: str):
the arguments are necessary for the definition
here
def get_country_samples(db: Session, skip: int = 0, limit: int = 100, endp_schema_key=''):
they are optional
I don't know how is better, we can make it like this
def get_country_samples(db: Session, endp_schema_key: str, skip: int = 0, limit: int = 100)

Comment (Mil-m) [reply kooplex-proxy/proxy/crud.py:26]: by this way db and endp_schema_key will be required parameters and skill with limit are optional

Comment (davidyuyuan) [reply kooplex-proxy/proxy/crud.py:26]: I think that the 2nd one is better.

**Review by davidyuyuan — CHANGES_REQUESTED** [review]

File: kooplex-proxy/proxy/crud.py (hunk 9 above)
Comment (davidyuyuan) [reply kooplex-proxy/proxy/crud.py:26]: Sorry, I should be clear. I think that "def get_country_samples(db: Session, endp_schema_key: str, skip: int = 0, limit: int = 100)" is better.

File: kooplex-proxy/proxy/main.py [hunk 10]
```diff
@@ -27,42 +27,42 @@ def read_root():
 
 
 @app.get("/country_samples/", response_model=List[schemas.MView_CountrySamples])
-def read_country_samples(skip: int = 0, limit: int = 100):
-    country_samples = crud.get_country_samples(next(get_db()), skip=skip, limit=limit)
+def read_country_samples(schema_key='', skip: int = 0, limit: int = 100):
```
Comment (davidyuyuan) [inline kooplex-proxy/proxy/main.py:30]: Are you planning to change schema_key='' to something similar to what you did in crud.py?

Comment (Mil-m) [reply kooplex-proxy/proxy/main.py:30]: is this case our schema_key will be required parameter for the endpoints without using the default schema name

Comment (davidyuyuan) [reply kooplex-proxy/proxy/main.py:30]: Oh I see. Please ignore my comments.

**Review by davidyuyuan — APPROVED** [review]

**Pull request merged**